iothub/device/src/IotHubBaseClient.cs

```diff
@@ -587,41 +533,9 @@ internal async Task OnMethodCalledAsync(DirectMethodRequest directMethodRequest)
                 return;
             }
 
-            Tuple<Func<DirectMethodRequest, object, Task<DirectMethodResponse>>, object> callbackContextPair = null;
             DirectMethodResponse directMethodResponse = null;
 
-            await _methodsSemaphore.WaitAsync().ConfigureAwait(false);
```

Comment:
The local variable saving a reference to the user supplied callback, I'm not seeing much value in having this semaphore here. Let me know if you feel there's a reason to have this semaphore here.
If we really wanted to ensure that there are no rtace conditions between a user setting this callback and us invoking it, we'll need to create a deep-copy of this callback-context pair.